Remora.Rest/Json/Internal/BoundDataObjectConverter.cs

```diff
@@ -0,0 +1,265 @@
+            bool allowsNull = property.AllowsNull();
+
+            var data = new DTOPropertyInfo
+            (
+                property,
+                readNames,
+                writeNames,
+                writer,
+                allowsNull,
+                defaultValue,
+                propertyOptions,
+                readProperties.Count
+            );
+
+            if (property.CanWrite)
+            {
+                // If a property is writable, it can be *read* from JSON.
+                readProperties.Add(data);
+            }
+
+            if (property.CanWrite || source.ShouldIncludeReadOnlyProperty(property))
+            {
+                // Any property that is writable and not excluded due to being read-only,
+                // can be *written* to JSON.
+                writeProperties.Add(data);
+            }
+        }
+
+        _writeProperties = writeProperties.ToArray();
+        _readProperties = readProperties.ToArray();
+
+        _readPropertiesByName = readProperties
+            .SelectMany(p => p.ReadNames.Select((n, i) => (IsPrimary: i == 0, n, DTOProperty: p)))
+            .ToDictionary(x => x.n, x => (x.IsPrimary, x.DTOProperty));
+    }
+
+    private static JsonSerializerOptions CreatePropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
+    {
+        if (converter == null)
+        {
+            return options;
+        }
+
+        var cloned = new JsonSerializerOptions(options);
+        cloned.Converters.Insert(0, converter);
+
+        return cloned;
+    }
+
+    /// <inheritdoc />
+    public override TInterface Read
+    (
+        ref Utf8JsonReader reader,
+        Type typeToConvert,
+        JsonSerializerOptions options
+    )
+    {
+        if (reader.TokenType != JsonTokenType.StartObject || !reader.Read())
+        {
+            throw new JsonException();
+        }
+
+        var readProperties = _readProperties;
+
+        // Possible Optimization: Cache this array and reuse it.
+        // F.e. in a thread-static field. This method isn't async anyways.
+
+        // Avoid creating a dictionary and fill the arguments array directly
+        var constructorArguments = new object?[readProperties.Length];
+
+        // We use "Missing" as a sentinel to indicate that a value hasn't
+        // been seen yet without needing to allocate additional memory.
+        constructorArguments.AsSpan().Fill(DataObjectConverterShared.Missing);
+
+        while (reader.TokenType != JsonTokenType.EndObject)
+        {
+            var propertyName = reader.GetString()!;
+            if (!reader.Read())
+            {
+                throw new JsonException();
+            }
+
+            var (isPrimaryChoice, dtoProperty) = GetReadPropertyInfo(propertyName);
+
+            if (dtoProperty is null)
+            {
+                if (!_allowExtraProperties)
+                {
+                    throw new JsonException();
```

Comment:
I would prefer here to include a message about the missing property in the exception.

Reply:
There was initially no message for this, but I agree.
Something like: `DTO disallows extra properties and has no matching property for JSON property {propertyName}.`

Reply:
I'm fine with that message.

Reply:
https://github.com/Remora/Remora.Rest/blob/37c3d450f268a39b3baae9976a071a09f2a5be7b/Remora.Rest/Json/Internal/BoundDataObjectConverter.cs#L177

Reply:
Looks good.